File: 1608-special-array-with-x-elements-greater-than-or-equal-x/1608-special-array-with-x-elements-greater-than-or-equal-x.py
class Solution:
    def specialArray(self, nums: List[int]) -> int:
        nums.sort()
        for i in range(len(nums) + 1):
            idx = bisect_left(nums, i, 0, len(nums))
            # print((len(nums) - idx))
            if((len(nums) - idx) == i):
                return i
        
        return -1

File: 1608-special-array-with-x-elements-greater-than-or-equal-x/1608-special-array-with-x-elements-greater-than-or-equal-x.go
func specialArray(nums []int) int {
    slices.Sort(nums)
    var binSearch = func (val int) int {
        var low, high = 0, len(nums) - 1
        for low <= high {
            var mid = low + ((high - low) / 2)
            if nums[mid] >= val {
                high = mid - 1
            } else {
                low = mid + 1
            }
        }
        return low
    }
    
    var left, right = 0, 1001
    for left <= right {
        var midVal = left + ((right - left) / 2)
        var binIndex = binSearch(midVal)
        if midVal == len(nums) - binIndex {
            return midVal
        }
        
        if midVal < len(nums) - binIndex {
            left = midVal + 1
        } else {
            right = midVal - 1
        }
    }
    
    return -1
}